function-sidecar/pkg/dispatcher/grpc/grpc.go

```diff
@@ -52,44 +59,100 @@ func (this *grpcDispatcher) Closed() <-chan struct{} {
 	return this.closed
 }
 
-func (this *grpcDispatcher) handleIncoming() {
-	for {
-		select {
-		case in, open := <-this.input:
-			if open {
-				grpcMessage := toGRPC(in)
-				err := this.stream.Send(grpcMessage)
-				if err != nil {
-					if streamClosureDiagnosed(err) {
-						close(this.closed)
-						return
-					}
-
-					log.Printf("Error sending message to function: %v", err)
-				}
+// A CorrelationStrategy groups together messages that belong to the same "spatial" stream (the "temporal" aspect
+// of grouping is taken care of by a WindowStrategy). Several invocations for different spatial groups may happen concurrently.
+type CorrelationStrategy func(message.Message) interface{}
+
+// A WindowingStrategyFactory is a function for creating a new instance of a (most certainly stateful) WindowingStrategy,
+// given a new, first message for a stream.
+type WindowingStrategyFactory func(message.Message) WindowingStrategy
+
+//
+type WindowingStrategy interface {
+	// ShouldClose notifies this strategy that a new message has just been sent to the function, resulting in the optional
+	// error passed as 2nd argument.
+	// ShouldClose should return true as its first result if this strategy decides that the current stream should end.
+	ShouldClose(in message.Message, err error) bool
```

Comment:
As defined, windows cannot consist of a single message since `ShouldClose` is called _after_ the message has been processed by the function. Why not remove the `Message` parameter from the `WindowingStrategyFactory` and then `ShouldClose` can be called with the first message **and** optional error? This allows the strategy to deal with an error from the first message and also allows the strategy to produce a window of size 1.

Reply:
I don't understand your comment. What we cannot do is windows of size _zero_, not one. Or am I missing something?

While not 100% necessary, passing the message to the factory is for strategies that need info from the first message (think time windows with _event time_ semantics)

Reply:
Oh, is the protocol that the first message is passed both to the factory and also to `ShouldClose`?  It would be good to document that if it's the case. If so, that would enable a window size of 1.

Reply:
Yes.

I thought you guys heard me say "window of size 1" enough times :) How could you even think it did not support that. I have recurring dreams of windows of size 1!